serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,282 @@
+	}
+	pod := &Pod{}
+	if err := c.do(req, pod); err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+// PatchPod updates the pod's tags to the given ones.
+// It does so non-destructively, or in other words, without tearing down
+// the pod.
+func (c *Client) PatchPod(namespace, podName string, patches ...*Patch) error {
+	endpoint := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s", namespace, podName)
+	method := http.MethodPatch
+
+	// Validate that we received required parameters.
+	if namespace == "" {
+		return ErrNamespaceUnset
+	}
+	if podName == "" {
+		return ErrPodNameUnset
+	}
+	if len(patches) == 0 {
+		// No work to perform.
+		return nil
+	}
+
+	var jsonPatches []interface{}
+	for _, patch := range patches {
+		if patch.Operation == Unset {
+			return errors.New("patch operation must be set")
+		}
+		jsonPatches = append(jsonPatches, map[string]interface{}{
+			"op":    patch.Operation.String(),
+			"path":  patch.Path,
+			"value": patch.Value,
+		})
+	}
+	body, err := json.Marshal(jsonPatches)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(method, c.config.Host+endpoint, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json-patch+json")
+	return c.do(req, nil)
+}
+
+// do executes the given request, retrying if necessary.
+func (c *Client) do(req *http.Request, ptrToReturnObj interface{}) error {
+	// Finish setting up a valid request.
+	req.Header.Set("Authorization", "Bearer "+c.config.BearerToken)
+	req.Header.Set("Accept", "application/json")
+	client := cleanhttp.DefaultClient()
+	client.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: c.config.CACertPool,
+		},
+	}
+
+	// Execute and retry the request. This exponential backoff comes
+	// with jitter already rolled in.
+	var lastErr error
+	b := backoff.NewExponentialBackOff()
+	for i := 0; i < maxRetries; i++ {
+		if i != 0 {
+			select {
+			case <-c.stopCh:
+				return nil
+			case <-time.NewTimer(b.NextBackOff()).C:
+				// Continue to the request.
+			}
+		}
+		shouldRetry, err := c.attemptRequest(client, req, ptrToReturnObj)
+		if !shouldRetry {
+			// The error may be nil or populated depending on whether the
+			// request was successful.
+			return err
+		}
+		lastErr = err
+	}
+	return lastErr
+}
+
+// attemptRequest tries one single request. It's in its own function so each
+// response body can be closed before returning, which would read awkwardly if
+// executed in a loop.
+func (c *Client) attemptRequest(client *http.Client, req *http.Request, ptrToReturnObj interface{}) (shouldRetry bool, err error) {
+	// Preserve the original request body so it can be viewed for debugging if needed.
+	// Reading it empties it, so we need to re-add it afterwards.
+	var reqBody []byte
+	if req.Body != nil {
+		reqBody, _ = ioutil.ReadAll(req.Body)
+		reqBodyReader := bytes.NewReader(reqBody)
+		req.Body = ioutil.NopCloser(reqBodyReader)
+	}
+
+	resp, err := client.Do(req)
```

Comment:
Could this panic if `resp` was nil but no error?  I doubt that would happen but curious what you think.

Reply:
Oh good question! Actually, one cool thing is that the Go `http.Client` guarantees that if the `err` is nil, the `resp` is always non-nil and it's the caller's duty to close the response body (or the caller may end up with a memory leak!) More [here](https://golang.org/pkg/net/http/#Client.Do). Thus, my habitual pattern when using it is some variation of:
```
resp, err := client.Do(req)
if err != nil {
	return false, err
}
defer resp.Body.Close() // Maybe log the error returned
```

Reply:
Thanks for the info!